src/Analyser/NodeScopeResolver.php

```diff
@@ -2585,7 +2585,11 @@ private function getPhpDocReturnType(?PhpDocBlock $phpDocBlock, ResolvedPhpDocBl
 			);
 		}
 
-		if (($phpDocBlock !== null && $phpDocBlock->isExplicit()) || $nativeReturnType->isSuperTypeOf($phpDocReturnType)->yes()) {
+		if ($phpDocBlock === null || $phpDocBlock->isExplicit()) {
+			return $phpDocReturnType;
+		}
+
+		if ($nativeReturnType->isSuperTypeOf(TemplateTypeHelper::resolveToBounds($phpDocReturnType))->yes()) {
```

Comment:
This is consistent with PhpFunctionReflection / TypehintHelper::decideTypeFromReflection